File: apps/desktop/src-tauri/src/application/conduct_service.rs
use std::sync::Arc;

use crate::domain::entities::{ConductConfig, ConductPointsLog, UserConductPoints};
use crate::domain::ports::ConductRepository;

pub struct ConductService {
    repo: Arc<dyn ConductRepository>,
}

impl ConductService {
    pub fn new(repo: Arc<dyn ConductRepository>) -> Self {
        Self { repo }
    }

    pub async fn get_config(&self, guild_id: String) -> Result<ConductConfig, String> {
        self.repo.get_config(guild_id).await
    }

    pub async fn get_leaderboard(&self, guild_id: String) -> Result<Vec<UserConductPoints>, String> {
        self.repo.get_leaderboard(guild_id).await
    }

    pub async fn get_points(&self, guild_id: String, user_id: String) -> Result<UserConductPoints, String> {
        self.repo.get_points(guild_id, user_id).await
    }

    pub async fn get_log(&self, guild_id: String, user_id: String) -> Result<Vec<ConductPointsLog>, String> {
        self.repo.get_log(guild_id, user_id).await
    }

    pub async fn adjust_points(&self, guild_id: String, user_id: String, amount: i32, reason: String) -> Result<UserConductPoints, String> {
        self.repo.adjust_points(guild_id, user_id, amount, reason).await
    }
}

File: apps/web/src/services/conductService.ts
import { httpGet, httpPost } from "@/api/http";
import type { ConductConfig, UserConductPoints, ConductPointsLog } from "@/types";

export interface SaveConductConfigPayload {
  guild_id: string;
  max_points?: number;
  regen_amount?: number;
  regen_interval?: string;
  penalty_warn?: number;
  penalty_delete?: number;
  penalty_mute?: number;
  penalty_ban?: number;
}

export const conductService = {
  getConfig(guildId: string): Promise<ConductConfig> {
    return httpGet(`/api/conduct/config/${guildId}`);
  },
  saveConfig(body: SaveConductConfigPayload): Promise<ConductConfig> {
    return httpPost(`/api/conduct/config`, body);
  },
  runRegenTick(): Promise<unknown> {
    return httpPost(`/api/conduct/regen-tick`);
  },
  syncBanProposals(): Promise<{ created: number }> {
    return httpPost(`/api/conduct/sync-ban-proposals`);
  },
  getLeaderboard(guildId: string): Promise<UserConductPoints[]> {
    return httpGet(`/api/conduct/${guildId}/leaderboard`);
  },
  getPoints(guildId: string, userId: string): Promise<UserConductPoints> {
    return httpGet(`/api/conduct/${guildId}/${userId}`);
  },
  getLog(guildId: string, userId: string): Promise<ConductPointsLog[]> {
    return httpGet(`/api/conduct/${guildId}/${userId}/log`);
  },
  adjustPoints(guildId: string, userId: string, amount: number, reason: string): Promise<unknown> {
    return httpPost(`/api/conduct/${guildId}/${userId}/add`, { amount, reason });
  },
};
